tests/h2-tests/tests/flow_control.rs

```diff
@@ -1561,3 +1561,86 @@ async fn data_padding() {
 
     join(srv, h2).await;
 }
+
+
+#[tokio::test]
+async fn notify_on_send_buffer_available() {
+    // This test ensures that the stream gets notified when there is additional
+    // send buffer space.
+    h2_support::trace_init!();
+
+    let (io, mut client) = mock::new();
+
+
+    let client = async move {
+        let settings = client.assert_server_handshake().await;
+        assert_default_settings!(settings);
+        client.send_frame(
+            frames::headers(1)
+                .request("GET", "https://www.example.com/")
+                .eos()
+        )
+        .await;
+        client.send_frame(
+            frames::headers(3)
+                .request("GET", "https://www.example.com/")
+                .eos()
+        )
+        .await;
+        client.recv_frame(frames::headers(1).response(200)).await;
+        client.recv_frame(frames::headers(3).response(200)).await;
+        dbg!(11);
+        client.recv_frame(frames::data(1, &b"abcde"[..]).eos()).await;
+        dbg!(31);
+        client.recv_frame(frames::data(3, &b"abcde"[..])).await;
+        dbg!(32);
+        client.recv_frame(frames::data(3, &b"abcde"[..])).await;
+        dbg!(33);
+        client.recv_frame(frames::data(3, &b"abcde"[..])).await;
+        dbg!(34);
+        client.recv_frame(frames::data(3, &b""[..]).eos()).await;
+    };
+
+    let srv = async move {
+        let mut srv = server::Builder::new()
+            .max_send_buffer_size(5)
+            .handshake::<_, Bytes>(io)
+            .await
+            .expect("handshake");
+
+        let (_req, mut reply1) = srv.next().await.unwrap().unwrap();
+        let (_req, mut reply2) = srv.next().await.unwrap().unwrap();
+
+        let mut stream1 = reply1.send_response(http::Response::new(()), false).unwrap();
+        let mut stream2 = reply2.send_response(http::Response::new(()), false).unwrap();
+        drop((reply1, reply2));
+
+        let t0 = tokio::spawn(async move {
+            assert!(srv.next().await.is_none(), "unexpected request");
+        });
+        let t1 = tokio::spawn(async move {
+            eprintln!("[t1] RESERVE 1 cap");
+            stream1.reserve_capacity(1);
+            stream1 = util::wait_for_capacity(stream1, 1).await;
+            eprintln!("[t1] got 1 cap");
+            stream1.send_data("abcde".into(), true).unwrap();
+        });
+        let t2 = tokio::spawn(async move {
+            for n in 0..3 {
+                eprintln!("[t2] RESERVE 1 cap, loop {}", n);
+                stream2.reserve_capacity(1);
+                stream2 = util::wait_for_capacity(stream2, 1).await;
+                eprintln!("[t2] got 1 cap, loop {}", n);
+                stream2.send_data("abcde".into(), false).unwrap();
+            }
+
+            stream2.send_data("".into(), true).unwrap();
+        });
+
+        t2.await.expect("srv body spawn");
+        t1.await.expect("srv body spawn");
+        t0.await.expect("srv end");
```

Comment:
AFAIK the test is still wrong. We await for `t2` before we await for `t0`, so `t0` is pending and nothing will wake it up when `t2` notifies the connection task that it has frames to send, so the send buffer size never decreases and everything gets stuck.

Reply:
Never mind I misread. That being said, the test really doesn't help much finding where the issue with my patch is :/

Reply:
https://github.com/hyperium/h2/blob/2c53d600989209e263c8157df6e6e2280a9a3355/tests/h2-support/src/util.rs#L56-L66

There, `poll_capacity` returns `Ready(Some(0))`, and `capacity` returns `0`, so `act >= self.target` is `false`, and `Poll::Pending` gets returned even though nothing has been set up to wake up the task.

What do you think is wrong here? That `poll_capacity` returns `Ready(Some(0))`, or that `wait_for_capacity` is badly conceived, or both?